### ISSUE TITLE
[Request] Add support for image exposure adjustment matrixcolor

### ISSUE THREAD

**Issue by yourherobby** [renpy/renpy#2517]
it would be very useful i think for making dynamic time of day scenes without need to re render same image several times

**Comment by renpytom** [comment]
I'm not opposed, certainly. But do you have any idea what this would actually mean?

**Comment by yourherobby** [comment]
i dont understand what your question. are you implying there is a reason that there you cant add exposure control? all i know if i could adjust exposure from within renpy i would have to worry less about one aspect of color correction that is very detrimental

**Comment by renpytom** [comment]
I guess I don't know what exposure control is, when it comes to a 2d drawn image. I'm curious if you have any information about, say, an exposure control filter?

**Comment by yourherobby** [comment]
ah ok i can give some references but i am not an expert so i cant give professional description. the way i understand it exposure would control the amount of light in an image. this works best with 16bit image, i am not sure if renpy supports this i havent checked but its useful even in 8bit for minor tweaks. here are some references:

https://en.wikipedia.org/wiki/Exposure_%28photography%29
https://docs.blender.org/manual/en/2.79/render/blender_render/world/exposure.html
https://affinity.help/photo/en-US.lproj/index.html?page=pages/Adjustments/adjustment_exposure.html?title=Exposure%20adjustment
https://docs.gimp.org/2.10/en/gimp-filter-exposure.html
https://www.gimp.org/tutorials/Blending_Exposures/

like i said i no expert but i will answer anything i can!

**Comment by yourherobby** [comment]
https://stackoverflow.com/questions/12166117/what-is-the-math-behind-exposure-adjustment-on-photoshop
https://photo.stackexchange.com/questions/27943/how-does-software-raise-exposure-in-post-processing

**Comment by yourherobby** [comment]
exposure is like brightness adjustment but with a different curve to shadows and highlights, i think.

but if you are going to delve into this, i suggest 2 more separate filters. one that targets shadows and other that targets highlights it is like exposure but it allows for separate adjustment of  both shadows and highlights.

https://daviesmediadesign.com/gimp-2-10-shadows-highlights-feature-text-version/

so, if renpy had these 3 filters. exposure, shadow and highlights. it would give people who use 3d renderings a huge benefit because as of now these corrections have to be done before using them in renpy and once you do them it is baked . these 3 are the most important in my opinion, i understand this is probably a lot of work in terms of research but i assure you that it would be useful as proper color correction is very important to the quality of pre rendered content. having ability to do those corrections from within renpy would open many possibilities like having different styles and tweaking the corrections over time something that might prove more difficult if you have a lot of content and then have to find the files and open them in software and then export to proper formats  and then have users re download everything etc.  and also with more careful adjustments and combinations even creating different time of days with much less files. 

bottom line is: opening new possibilities, easing the 3d workflow and making it more forgiving, saving a ton of disk space and time in pre-baking color corrections that can(theoretically) be comfortably done from within game engine(i assume)

also if you feel generous a sharpen filter would be amazing. and even with that it would be 4 new filters but if renpy had that it would be on whole new level of usefulness in my opinion

**Comment by yourherobby** [comment]
here is a random 3d character i found that doesnt have any obvious corrections, the one to the left has no corrections one to the right has stronger shadows and stronger highlights effects is that it looks less flat. how much someone corrects this is a matter of taste i suppose i overblown it here a bit to make it more obvious but the point is you have to do it in image editing software and if later on i feel that maybe effect was too strong then i have to find that file and redo part of the workflow all again. if this was in renpy it would be a matter of tweaking a variable. so yeah, idk if your matrixcolor thing is designed to handle these kind of filters but if yes again it would be useful. just wanted to actually show you what kind of difference it makes.

![Untitled](https://user-images.githubusercontent.com/75397087/102927831-ff5d7f00-4497-11eb-89c4-b98ca9a52a91.jpg)

**Comment by Gouvernathor** [comment]
This kind of correction is used in video editing, to tweak the colors. Typically it's done by applying a kind of [s-curve](https://easings.net/#easeInOutQuint) to the color channel(s). As this is not a linear transformation, it can't be done as such with matrices. Trying to approach it using affine transformations would result in burned white and closed black areas where there had been nuances.

However, you could probably do it through blood and tears with the new [post-processing](https://www.renpy.org/dev-doc/html/model.html).
